### PR TITLE
Implemented multiple reports page

### PR DESCRIPTION
**Pull request by Sandipan-Ghosh**
**Fixed Issues**

21977, 21978, 21979, 21980, 21982, 21983, 21984, 21985, 21986, 21987, 21988, 21990, 21991,21992, 21993

**Files Changed**

* IReportRepository

* ReportRepository

* ReportController

* Report.cs

* TestAttendee.cs

* TestStatus.cs

* testAttendee.ts

* report.ts

* test-report.component.ts

* reports.routing.ts

* enum-test-state.ts

* startup.cs

* test-report.html

* question-programming.html

* report.services.ts

* report.module.ts

* test-dashboard.html

**Checks**

- [x] Naming Conventions
- [x] Server side code comments (proper English, grammar and no spelling mistakes)
- [x] Client side code comments (proper English, grammar and no spelling mistakes)
- [x] Unused variables, methods, blank spaces and name spaces. 
- [x] Optimal Code and code formatting
- [x] Commit History is proper, no merge is done. 
- [x] Commit/pull request messages should be proper to justify your feature
- [x] All test cases are executed provided by tester.

**Known Issues**

**Comments**

---
This change is [Reviewable](https://reviewable.io/reviews/promact/trappist/342)

### CODE REVIEW

File: Trappist/src/Promact.Trappist.Repository/Reports/IReportRepository.cs [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+namespace Promact.Trappist.Repository.Reports
+{
+    public interface IReportRepository
+    {
+        /// <summary>
+        /// Method to get the details of all the Test-Attendees
+        /// </summary>
+        /// <param name="id">Id of the respective Test</param>
+        /// <returns></returns>
+        Task<ICollection<TestAttendees>> GetAllTestAttendeesAsync(int id);
```
Comment (Krunal-Promact) [inline Trappist/src/Promact.Trappist.Repository/Reports/IReportRepository.cs]: Please use IEnumerable

File: Trappist/src/Promact.Trappist.Repository/Reports/ReportRepository.cs [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+
+        #region Public Method
+        public async Task<ICollection<TestAttendees>> GetAllTestAttendeesAsync(int id)
+        {
+            return await _dbContext.TestAttendees.Where(t => t.TestId == id).Include(x => x.Report).ToListAsync();
+        }
+
+        public async Task SetStarredCandidateAsync(int id)
+        {
+            var studentToBeStarred = await _dbContext.TestAttendees.Where(x => x.Id == id).FirstOrDefaultAsync();
```
Comment (Krunal-Promact) [inline Trappist/src/Promact.Trappist.Repository/Reports/ReportRepository.cs]: primary key search should be implemented using Find method.

File: Trappist/src/Promact.Trappist.Web/wwwroot/app/reports/reports.routing.ts [hunk 3]
```diff
@@ -7,11 +7,11 @@ import { IndividualReportComponent } from './individual-report/individual-report
 
 const reportsRoutes: Routes = [
     {
-        path: 'reports',
+        path: 'tests',
```
Comment (Krunal-Promact) [inline Trappist/src/Promact.Trappist.Web/wwwroot/app/reports/reports.routing.ts]: path with tests already exist. why are you creating a duplicate?

File: Trappist/src/Promact.Trappist.Web/wwwroot/app/reports/test-report/test-report.component.ts [hunk 4]
```diff
@@ -1,33 +1,211 @@
...
+        return this.count = this.testAttendeeArray.length;
+    }
+
+    /**
+     * Keeps tarck of starting index of the active page
+     * @param activePage
+     * @param count
+     */
+    startingIndexOfActivePage(activePage: number, count: number) {
+
```
Comment (Krunal-Promact) [inline Trappist/src/Promact.Trappist.Web/wwwroot/app/reports/test-report/test-report.component.ts]: what is the meaning of this extra space?

File: Trappist/src/Promact.Trappist.Repository/Reports/ReportRepository.cs [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+            studentToBeStarred.StarredCandidate = studentToBeStarred.StarredCandidate ? false : true;
```
Comment (chintans) [inline Trappist/src/Promact.Trappist.Repository/Reports/ReportRepository.cs]: MINOR Remove the unnecessary Boolean literal(s). [rule](http://server12:9005/coding_rules#rule_key=csharpsquid%3AS1125)

**Comment by chintans** [comment]
SonarQube analysis reported 2 issues
* ![MINOR][MINOR] 2 minor

Watch the comments in this conversation to review them.

[MINOR]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png 'Severity: MINOR'

File: Trappist/src/Promact.Trappist.Repository/Reports/ReportRepository.cs [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+        public async Task<IEnumerable<TestAttendees>> GetAllTestAttendeesAsync(int id)
+        {
+            var result = await _dbContext.TestAttendees.Where(t => t.TestId == id).Include(x => x.Report).ToListAsync();
+            return result;
+        }
+
+        public async Task SetStarredCandidateAsync(int id)
+        {
+            var studentToBeStarred = await _dbContext.TestAttendees.FindAsync(id);
+            studentToBeStarred.StarredCandidate = studentToBeStarred.StarredCandidate ? false : true;
```
Comment (chintans) [inline Trappist/src/Promact.Trappist.Repository/Reports/ReportRepository.cs]: MINOR Remove the unnecessary Boolean literal(s). [rule](http://server12:9005/coding_rules#rule_key=csharpsquid%3AS1125)

**Comment by chintans** [comment]
SonarQube analysis reported 1 issue
* ![MINOR][MINOR] 1 minor

Watch the comments in this conversation to review them.

[MINOR]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png 'Severity: MINOR'

File: Trappist/src/Promact.Trappist.Repository/Reports/ReportRepository.cs (hunk 6 above)
Comment (chintans) [inline Trappist/src/Promact.Trappist.Repository/Reports/ReportRepository.cs]: MINOR Remove the unnecessary Boolean literal(s). [rule](http://server12:9005/coding_rules#rule_key=csharpsquid%3AS1125)

**Comment by chintans** [comment]
SonarQube analysis reported 1 issue
* ![MINOR][MINOR] 1 minor

Watch the comments in this conversation to review them.

[MINOR]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png 'Severity: MINOR'

File: Trappist/src/Promact.Trappist.Repository/Reports/ReportRepository.cs (hunk 6 above)
Comment (chintans) [inline Trappist/src/Promact.Trappist.Repository/Reports/ReportRepository.cs]: MINOR Remove the unnecessary Boolean literal(s). [rule](http://server12:9005/coding_rules#rule_key=csharpsquid%3AS1125)

**Comment by chintans** [comment]
SonarQube analysis reported 1 issue
* ![MINOR][MINOR] 1 minor

Watch the comments in this conversation to review them.

[MINOR]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png 'Severity: MINOR'

File: Trappist/src/Promact.Trappist.Web/wwwroot/app/reports/test-report/test-report.component.ts [hunk 7]
```diff
@@ -1,33 +1,205 @@
...
+     * @param testId
+     */
+    setAllCandidatesStarred() {
+        let status = this.headerStarStatus === 'star_border' ? true : false;
+        this.reportService.setAllCandidatesStarred(status, this.searchString, this.selectedTestStatus).subscribe(
+            response => {
+                this.testAttendeeArray.forEach(k => k.starredCandidate = status);
+                [this.headerStarStatus, this.isAllCandidateStarred] = status ? ['star', true] : ['star_border', false];
+            },
+            err => {
```
Comment (Krunal-Promact) [inline Trappist/src/Promact.Trappist.Web/wwwroot/app/reports/test-report/test-report.component.ts]: why is error handling empty?

File: Trappist/src/Promact.Trappist.Web/wwwroot/app/reports/test-report/test-report.component.ts [hunk 8]
```diff
@@ -1,33 +1,205 @@
...
+     * Toggles the star condition of a candidate
+     * @param testAttendee
+     */
+    setStarredCandidate(testAttendee: TestAttendee) {
+        this.reportService.setStarredCandidate(testAttendee.id).subscribe(
+            response => {
+                this.testAttendeeArray.find(x => x.id === testAttendee.id).starredCandidate = !testAttendee.starredCandidate;
+                [this.headerStarStatus, this.isAllCandidateStarred] = this.testAttendeeArray.some(x => !x.starredCandidate) ? ['star_border', false] : ['star', true];
+            },
+            err => {
```
Comment (Krunal-Promact) [inline Trappist/src/Promact.Trappist.Web/wwwroot/app/reports/test-report/test-report.component.ts]: same as above

File: Trappist/src/Promact.Trappist.Web/wwwroot/app/reports/testAttendee.ts [hunk 9]
```diff
@@ -0,0 +1,14 @@
+﻿import { Report } from '../reports/report';
+export class TestAttendee {
```
Comment (Krunal-Promact) [inline Trappist/src/Promact.Trappist.Web/wwwroot/app/reports/testAttendee.ts]: extra space should be there between import and export

File: Trappist/src/Promact.Trappist.Repository/Reports/ReportRepository.cs (hunk 6 above)
Comment (chintans) [inline Trappist/src/Promact.Trappist.Repository/Reports/ReportRepository.cs]: MINOR Remove the unnecessary Boolean literal(s). [rule](http://server12:9005/coding_rules#rule_key=csharpsquid%3AS1125)

**Comment by chintans** [comment]
SonarQube analysis reported 1 issue
* ![MINOR][MINOR] 1 minor

Watch the comments in this conversation to review them.

[MINOR]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png 'Severity: MINOR'

File: Trappist/src/Promact.Trappist.Repository/Reports/ReportRepository.cs [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+        public async Task<IEnumerable<TestAttendees>> GetAllTestAttendeesAsync(int id)
+        {
+            return await _dbContext.TestAttendees.Where(t => t.TestId == id).Include(x => x.Report).ToListAsync();
+           
+        }
+
+        public async Task SetStarredCandidateAsync(int id)
+        {
+            var studentToBeStarred = await _dbContext.TestAttendees.FindAsync(id);
+            studentToBeStarred.StarredCandidate = studentToBeStarred.StarredCandidate ? false : true;
```
Comment (chintans) [inline Trappist/src/Promact.Trappist.Repository/Reports/ReportRepository.cs]: MINOR Remove the unnecessary Boolean literal(s). [rule](http://server12:9005/coding_rules#rule_key=csharpsquid%3AS1125)

**Comment by chintans** [comment]
SonarQube analysis reported 1 issue
* ![MINOR][MINOR] 1 minor

Watch the comments in this conversation to review them.

[MINOR]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png 'Severity: MINOR'

File: Trappist/src/Promact.Trappist.Repository/Reports/IReportRepository.cs [hunk 11]
```diff
@@ -0,0 +1,32 @@
...
+        Task SetStarredCandidateAsync(int id);
+
+        /// <summary>
+        /// Method to set all candidates with matching criteria as starred candidate
+        /// </summary>
+        /// <param name="status">Star status of the candidates</param>
+        /// <param name="selectedTestStatus">Test end status of the candidates</param>
+        /// <param name="searchString">The search string provided by the user</param>
+        /// <returns>Star status of the candidates</returns>
+        Task SetAllCandidateStarredAsync(bool status,  int selectedTestStatus, string SearchString);
```
Comment (Krunal-Promact) [inline Trappist/src/Promact.Trappist.Repository/Reports/IReportRepository.cs]: formatting issue. Please use ctrl K + ctrl D

File: Trappist/src/Promact.Trappist.Repository/Reports/ReportRepository.cs [hunk 12]
```diff
@@ -0,0 +1,57 @@
...
+        {
+            _dbContext = dbContext;
+        }
+        #endregion
+
+        #region Public Method
+        public async Task<IEnumerable<TestAttendees>> GetAllTestAttendeesAsync(int id)
+        {
+            return await _dbContext.TestAttendees.Where(t => t.TestId == id).Include(x => x.Report).ToListAsync();
+
```
Comment (Krunal-Promact) [inline Trappist/src/Promact.Trappist.Repository/Reports/ReportRepository.cs]: extra space.

File: Trappist/src/Promact.Trappist.Repository/Reports/ReportRepository.cs [hunk 13]
```diff
@@ -0,0 +1,57 @@
...
+        #region Public Method
+        public async Task<IEnumerable<TestAttendees>> GetAllTestAttendeesAsync(int id)
+        {
+            return await _dbContext.TestAttendees.Where(t => t.TestId == id).Include(x => x.Report).ToListAsync();
+           
+        }
+
+        public async Task SetStarredCandidateAsync(int id)
+        {
+            var studentToBeStarred = await _dbContext.TestAttendees.FindAsync(id);
```
Comment (Krunal-Promact) [inline Trappist/src/Promact.Trappist.Repository/Reports/ReportRepository.cs]: what if a record with this id does not exist in the database?

File: Trappist/src/Promact.Trappist.Test/Reports/ReportsRepositoryTest.cs [hunk 14]
```diff
@@ -1,13 +1,131 @@
...
+
+        /// <summary>
+        ///This method used to initialize test attendee model parameters.  
+        /// </summary>
+        /// <returns>It return TestAttendee model object which contain first name,last name,email,contact number,roll number</returns>
+        private TestAttendees TestAttendeeReport(int id)
+        {
+            var testAttendee = new TestAttendees()
+            {
+                Id=1,
```
Comment (Krunal-Promact) [inline Trappist/src/Promact.Trappist.Test/Reports/ReportsRepositoryTest.cs]: formatting issue

File: Trappist/src/Promact.Trappist.Web/wwwroot/app/reports/report.service.ts [hunk 15]
```diff
@@ -0,0 +1,21 @@
...
+export class ReportService {
+    private reportsApiUrl = 'api/report';
+
+    constructor(private httpService: HttpService) {
+    }
+
+    getAllTestAttendees(testId: number) {
+        return this.httpService.get(this.reportsApiUrl+'/'+testId);
+    }
+    setStarredCandidate(attendeeId: number) {
```
Comment (Krunal-Promact) [inline Trappist/src/Promact.Trappist.Web/wwwroot/app/reports/report.service.ts]: please add a space after each method body.

File: Trappist/src/Promact.Trappist.Repository/Reports/ReportRepository.cs (hunk 13 above)
Comment (Sandipan-Ghosh) [inline Trappist/src/Promact.Trappist.Repository/Reports/ReportRepository.cs]: this id is the id of the student who is displayed on the multiple reports page, if it doesnot exist in the database then it wont be displayed in the reports page.

File: Trappist/src/Promact.Trappist.Web/wwwroot/app/reports/test-report/test-report.component.ts [hunk 16]
```diff
@@ -1,33 +1,215 @@
...
+    getTestAttendeeMatchingSearchCriteria(searchString: string) {
+        this.searchString = searchString;
+        this.filterList();
+    }
+
+    /**
+     * To determine whether search field will visible or hidden
+     */
+    showStatus() {
+        if (this.searchString.length > 0)
```
Comment (Krunal-Promact) [inline Trappist/src/Promact.Trappist.Web/wwwroot/app/reports/test-report/test-report.component.ts]: please optimize the code

File: Trappist/src/Promact.Trappist.Web/wwwroot/app/questions/questions-dashboard/questions-dashboard.component.ts [hunk 17]
```diff
@@ -131,6 +131,13 @@ export class QuestionsDashboardComponent implements OnInit {
         }
     }
 
+    /**
+     * To determine whether search field will be visible or hidden
+     */
+    showStatus() {
+        return this.showSearchInput = this.matchString.length > 0 ? true : false;
```
Comment (Krunal-Promact) [inline Trappist/src/Promact.Trappist.Web/wwwroot/app/questions/questions-dashboard/questions-dashboard.component.ts]: return this.showSearchInput = this.matchString.length > 0;

File: Trappist/src/Promact.Trappist.Web/wwwroot/app/reports/test-report/test-report.component.ts [hunk 18]
```diff
@@ -1,33 +1,212 @@
...
+            this.loader = false;
+        });
+    }
+
+    /**
+     * Sets all the candidates as starred candidates
+     * @param testId
+     */
+    setAllCandidatesStarred() {
+        let status = this.headerStarStatus === 'star_border' ? true : false;
```
Comment (Krunal-Promact) [inline Trappist/src/Promact.Trappist.Web/wwwroot/app/reports/test-report/test-report.component.ts]: let status = this.headerStarStatus === 'star_border'

File: Trappist/src/Promact.Trappist.Web/wwwroot/app/reports/test-report/test-report.component.ts [hunk 19]
```diff
@@ -1,33 +1,212 @@
...
+    getTestAttendeeMatchingSearchCriteria(searchString: string) {
+        this.searchString = searchString;
+        this.filterList();
+    }
+
+    /**
+     * To determine whether search field will be visible or hidden
+     */
+    showStatus() {
+        return this.showSearchInput = this.searchString.length > 0 ? true : false;
```
Comment (Krunal-Promact) [inline Trappist/src/Promact.Trappist.Web/wwwroot/app/reports/test-report/test-report.component.ts]: return this.showSearchInput = this.searchString.length > 0

File: Trappist/src/Promact.Trappist.Repository/Reports/ReportRepository.cs [hunk 20]
```diff
@@ -0,0 +1,75 @@
...
+            List <TestAttendees> attendeeList;
+
+            if (searchString == null)
+            {
+                searchString = "";
+            }
+           
+            if ((TestStatus)selectedTestStatus == TestStatus.AllCandidates)
+            {
+                attendeeList = await _dbContext.TestAttendees.Where(x => x.FirstName.Contains(searchString) || x.FirstName.Contains(searchString) || x.Email.Contains(searchString)).ToListAsync();
```
Comment (chintans) [inline Trappist/src/Promact.Trappist.Repository/Reports/ReportRepository.cs]: CRITICAL Identical sub-expressions on both sides of operator '||'. [rule](http://server12:9005/coding_rules#rule_key=csharpsquid%3AS1764)

**Comment by chintans** [comment]
SonarQube analysis reported 1 issue
* ![CRITICAL][CRITICAL] 1 critical

Watch the comments in this conversation to review them.

[CRITICAL]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-critical.png 'Severity: CRITICAL'

File: Trappist/src/Promact.Trappist.DomainModel/Enum/TestStatus.cs [hunk 21]
```diff
@@ -0,0 +1,10 @@
+﻿namespace Promact.Trappist.DomainModel.Enum
+{
```
Comment (iroshni) [inline Trappist/src/Promact.Trappist.DomainModel/Enum/TestStatus.cs]: does not seems formatted. Please press ctrl + d on this file.

File: Trappist/src/Promact.Trappist.DomainModel/Models/Report/Report.cs [hunk 22]
```diff
@@ -0,0 +1,28 @@
...
+using Promact.Trappist.DomainModel.Models.TestConduct;
+using System.ComponentModel.DataAnnotations;
+using System.ComponentModel.DataAnnotations.Schema;
+
+namespace Promact.Trappist.DomainModel.Models.Report
+{
+    public class Report : BaseModel
+    {
+        [Required]
+        public double TotalMarksScored { get; set; }
```
Comment (iroshni) [inline Trappist/src/Promact.Trappist.DomainModel/Models/Report/Report.cs]: reason for making it require? double/int/float datatypes are not nullable by default. it takes default values if not provided explicitly. So wouldn't make difference giving required attribute.

File: Trappist/src/Promact.Trappist.Repository/Reports/ReportRepository.cs [hunk 23]
```diff
@@ -0,0 +1,75 @@
...
+            studentToBeStarred.StarredCandidate = !studentToBeStarred.StarredCandidate;
+            await _dbContext.SaveChangesAsync();
+        }
+
+        public async Task SetAllCandidateStarredAsync(bool status, int selectedTestStatus, string searchString)
+        {
+            List <TestAttendees> attendeeList;
+
+            if (searchString == null)
+            {
```
Comment (iroshni) [inline Trappist/src/Promact.Trappist.Repository/Reports/ReportRepository.cs]: In if condition, curley braces can be eliminated if only one like needs to be exceucted in it. helps in making compact code.

File: Trappist/src/Promact.Trappist.Repository/Reports/ReportRepository.cs [hunk 24]
```diff
@@ -0,0 +1,75 @@
...
+                searchString = "";
+            }
+           
+            if ((TestStatus)selectedTestStatus == TestStatus.AllCandidates)
+            {
+                attendeeList = await _dbContext.TestAttendees.Where(x => x.FirstName.Contains(searchString) || x.LastName.Contains(searchString) || x.Email.Contains(searchString)).ToListAsync();
```
Comment (iroshni) [inline Trappist/src/Promact.Trappist.Repository/Reports/ReportRepository.cs]: We need to do search ignoring case. Please make sure the same is achieved.

File: Trappist/src/Promact.Trappist.Repository/Reports/ReportRepository.cs [hunk 25]
```diff
@@ -0,0 +1,75 @@
...
+            }
+
+            else
+            {
+                attendeeList = await _dbContext.TestAttendees.Include(x => x.Report).Where(x => x.Report.TestStatus == (TestStatus)selectedTestStatus
+                && (x.FirstName.Contains(searchString)|| x.LastName.Contains(searchString)|| x.Email.Contains(searchString))).ToListAsync();
```
Comment (iroshni) [inline Trappist/src/Promact.Trappist.Repository/Reports/ReportRepository.cs]: same as above

File: Trappist/src/Promact.Trappist.Web/wwwroot/app/reports/test-report/test-report.component.ts [hunk 26]
```diff
@@ -1,33 +1,229 @@
...
+    starredCandidateIdList: number[];
+    isAllCandidateStarred: boolean;
+    showStarCandidates: boolean;
+    loader: boolean;
+    test: Test;
+
+    constructor(private reportService: ReportService, private route: ActivatedRoute) {
+        this.testAttendeeArray = new Array<TestAttendee>();
+        this.attendeeArray = new Array<TestAttendee>();
+        this.typeOfTest = '0';
```
Comment (iroshni) [inline Trappist/src/Promact.Trappist.Web/wwwroot/app/reports/test-report/test-report.component.ts]: What does typeOfTest variable means ?

Comment (Sandipan-Ghosh) [inline Trappist/src/Promact.Trappist.Web/wwwroot/app/reports/test-report/test-report.component.ts]: type of test given by the candidate i.e. completed test, blocked or expired.

Comment (iroshni) [inline Trappist/src/Promact.Trappist.Web/wwwroot/app/reports/test-report/test-report.component.ts]: does this name justify what you want to achieve?

File: Trappist/src/Promact.Trappist.Web/Migrations/20170505052950_UpdatedReportsTable.cs [hunk 27]
```diff
@@ -0,0 +1,19 @@
+﻿using System;
+using System.Collections.Generic;
+using Microsoft.EntityFrameworkCore.Migrations;
+
+namespace Promact.Trappist.Web.Migrations
+{
+    public partial class UpdatedReportsTable : Migration
+    {
+        protected override void Up(MigrationBuilder migrationBuilder)
```
Comment (chintans) [inline Trappist/src/Promact.Trappist.Web/Migrations/20170505052950_UpdatedReportsTable.cs]: MAJOR Add a nested comment explaining why this method is empty, throw a 'NotSupportedException' or complete the implementation. [rule](http://server12:9005/coding_rules#rule_key=csharpsquid%3AS1186)

File: Trappist/src/Promact.Trappist.Web/Migrations/20170505052950_UpdatedReportsTable.cs [hunk 28]
```diff
@@ -0,0 +1,19 @@
...
+        {
+
+        }
+
+        protected override void Down(MigrationBuilder migrationBuilder)
```
Comment (chintans) [inline Trappist/src/Promact.Trappist.Web/Migrations/20170505052950_UpdatedReportsTable.cs]: MAJOR Add a nested comment explaining why this method is empty, throw a 'NotSupportedException' or complete the implementation. [rule](http://server12:9005/coding_rules#rule_key=csharpsquid%3AS1186)

File: Trappist/src/Promact.Trappist.Repository/Reports/ReportRepository.cs [hunk 29]
```diff
@@ -0,0 +1,71 @@
...
+        }
+
+        public async Task SetAllCandidateStarredAsync(bool status, int selectedTestStatus, string searchString)
+        {
+            List<TestAttendees> attendeeList;
+
+            if (searchString == null)
+                searchString = "";
+            else
+                searchString = searchString.ToLower();
```
Comment (chintans) [inline Trappist/src/Promact.Trappist.Repository/Reports/ReportRepository.cs]: MAJOR Define the locale to be used in this string operation. [rule](http://server12:9005/coding_rules#rule_key=csharpsquid%3AS1449)

**Comment by chintans** [comment]
SonarQube analysis reported 3 issues
* ![MAJOR][MAJOR] 3 major

Watch the comments in this conversation to review them.

[MAJOR]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR'

File: Trappist/src/Promact.Trappist.Core/Controllers/ReportController.cs [hunk 30]
```diff
@@ -0,0 +1,76 @@
...
+                return NotFound();
+
+            await _reportRepository.SetStarredCandidateAsync(attendeeId);
+            return Ok(attendeeId);
+        }
+
+        /// <summary>
+        /// Method to get test name
+        /// </summary>
+        /// <param name="Id"></param>
```
Comment (Krunal-Promact) [inline Trappist/src/Promact.Trappist.Core/Controllers/ReportController.cs]: description for id is missing.

File: Trappist/src/Promact.Trappist.Repository/Reports/IReportRepository.cs [hunk 31]
```diff
@@ -0,0 +1,47 @@
...
+        /// <param name="selectedTestStatus">Test end status of the candidates</param>
+        /// <param name="searchString">The search string provided by the user</param>
+        /// <returns>Star status of the candidates</returns>
+        Task SetAllCandidateStarredAsync(bool status, int selectedTestStatus, string SearchString);
+
+        /// <summary>
+        /// Method to check whether a candidate exist or not
+        /// </summary>
+        /// <param name="attendeeId">Id of the candidate</param>
+        /// <returns></returns>
```
Comment (Krunal-Promact) [inline Trappist/src/Promact.Trappist.Repository/Reports/IReportRepository.cs]: description of the return statement is missing.

File: Trappist/src/Promact.Trappist.Repository/Reports/IReportRepository.cs [hunk 32]
```diff
@@ -0,0 +1,47 @@
...
+        Task <bool> IsCandidateExistAsync(int attendeeId);
+        
+        /// <summary>
+        ///Method to get test name 
+        /// </summary>
+        /// <param name="id"></param>
```
Comment (Krunal-Promact) [inline Trappist/src/Promact.Trappist.Repository/Reports/IReportRepository.cs]: same description of id and return is missing.

File: Trappist/src/Promact.Trappist.Repository/Reports/ReportRepository.cs [hunk 33]
```diff
@@ -0,0 +1,71 @@
...
+        {
+            List<TestAttendees> attendeeList;
+
+            if (searchString == null)
+                searchString = "";
+            else
+                searchString = searchString.ToLowerInvariant();
+
+            if ((TestStatus)selectedTestStatus == TestStatus.AllCandidates)
+                attendeeList = await _dbContext.TestAttendees.Where(x => x.FirstName.ToLower().Contains(searchString) || x.LastName.ToLower().Contains(searchString) || x.Email.Contains(searchString)).ToListAsync();
```
Comment (Krunal-Promact) [inline Trappist/src/Promact.Trappist.Repository/Reports/ReportRepository.cs]: searchstring should also be converted to lower case

File: Trappist/src/Promact.Trappist.Repository/Reports/ReportRepository.cs [hunk 34]
```diff
@@ -0,0 +1,71 @@
...
+
+            else
+                attendeeList = await _dbContext.TestAttendees.Include(x => x.Report).Where(x => x.Report.TestStatus == (TestStatus)selectedTestStatus
+                && (x.FirstName.ToLower().Contains(searchString) || x.LastName.ToLower().Contains(searchString) || x.Email.Contains(searchString))).ToListAsync();
```
Comment (Krunal-Promact) [inline Trappist/src/Promact.Trappist.Repository/Reports/ReportRepository.cs]: same as above.

File: Trappist/src/Promact.Trappist.Web/wwwroot/app/reports/test-report/test-report.component.ts [hunk 35]
```diff
@@ -1,33 +1,229 @@
...
+        starAttendeeArray.forEach(x => {
+            if (x.report.testStatus === selectedTestStatus) {
+                tempAttendeeArray.push(x);
+            }
+        });
+
+        tempAttendeeArray.forEach(x => {
+            {
+                if (x.email.toLowerCase().includes(searchString)
+                    || x.firstName.toLowerCase().includes(searchString)
```
Comment (Krunal-Promact) [inline Trappist/src/Promact.Trappist.Web/wwwroot/app/reports/test-report/test-report.component.ts]: searchstring should also be converted to lower case

**Pull request merged**